File: view/admin/function/process_pdf.py
import fitz  # PyMuPDF
import mysql.connector
import re
import sys
import os

# รับ path ของ PDF จาก command line arguments
pdf_path = sys.argv[1]

# เชื่อมต่อกับฐานข้อมูล MySQL
conn = mysql.connector.connect(
    host='localhost',
    user='root',
    password='',
    database='salary_slip'
)
cursor = conn.cursor()

# เปิดไฟล์ PDF
pdf_document = fitz.open(pdf_path)
permanent_folder = "permanent_pdfs"
os.makedirs(permanent_folder, exist_ok=True)

# อ่านแต่ละหน้าใน PDF
for page_num in range(pdf_document.page_count):
    page = pdf_document.load_page(page_num)
    page_height = page.rect.height

    # แบ่งข้อมูลครึ่งบนและครึ่งล่างของหน้า
    top_half = page.get_text("text", clip=fitz.Rect(0, 0, page.rect.width, page_height / 2))
    bottom_half = page.get_text("text", clip=fitz.Rect(0, page_height / 2, page.rect.width, page_height))

    # สร้างไฟล์ PDF สำหรับครึ่งบนและครึ่งล่าง
    top_half_pdf = fitz.open()
    bottom_half_pdf = fitz.open()

    # ครึ่งบน
    top_rect = fitz.Rect(0, 0, page.rect.width, page_height / 2)
    top_page = pdf_document.load_page(page_num)
    top_page.set_cropbox(top_rect)
    top_half_pdf.insert_pdf(pdf_document, from_page=page_num, to_page=page_num)

    # ครึ่งล่าง
    bottom_rect = fitz.Rect(0, page_height / 2, page.rect.width, page_height)
    bottom_page = pdf_document.load_page(page_num)
    bottom_page.set_cropbox(bottom_rect)
    bottom_half_pdf.insert_pdf(pdf_document, from_page=page_num, to_page=page_num)

    # กำหนด path สำหรับบันทึกไฟล์ถาวร
    top_pdf_path = os.path.join(permanent_folder, f"top_{page_num + 1}.pdf")
    bottom_pdf_path = os.path.join(permanent_folder, f"bottom_{page_num + 1}.pdf")

    # บันทึกไฟล์ PDF ถาวร
    top_half_pdf.save(top_pdf_path)
    bottom_half_pdf.save(bottom_pdf_path)

    def extract_emp_id(text):
        # ใช้ regex เพื่อค้นหาเลขพนักงานที่ขึ้นต้นด้วย "P" และตามด้วยตัวเลข
        match = re.search(r'\bP\d{4,}\b', text)  # เลขพนักงานที่มี 4 ตัวอักษรขึ้นไป
        if match:
            return match.group(0)
        
        # ถ้าไม่พบในรูปแบบข้างต้น ลองหาโดยดูบริบทเพิ่มเติม (เช่นคำว่า "รหัสพนักงาน" หรือ "Emp")
        match = re.search(r'(รหัสพนักงาน|Emp|Employee ID)\s*[:\-]?\s*(P\d+)', text, re.IGNORECASE)
        if match:
            return match.group(2)

        # หากยังไม่พบ ให้ลองค้นหาด้วยรูปแบบอื่น ๆ ตามที่เห็นในไฟล์
        lines = text.split("\n")
        for line in lines:
            if re.match(r'P\d{4,}', line.strip()):
                return line.strip()

        # หากไม่พบเลขพนักงาน ให้คืนค่าเป็น "unknown"
        return 'unknown'

    # แยก emp_id สำหรับครึ่งบนและครึ่งล่าง
    emp_id_top = extract_emp_id(top_half)
    emp_id_bottom = extract_emp_id(bottom_half)

    # บันทึกข้อมูลในครึ่งบนพร้อม path ของไฟล์ PDF
    sql_top = "INSERT INTO tb_payslips (emp_id, page_number, part, content, slip_divide) VALUES (%s, %s, %s, %s, %s)"
    cursor.execute(sql_top, (emp_id_top, page_num + 1, 'top', top_half, top_pdf_path))
    conn.commit()

    # บันทึกข้อมูลในครึ่งล่างพร้อม path ของไฟล์ PDF
    sql_bottom = "INSERT INTO tb_payslips (emp_id, page_number, part, content, slip_divide) VALUES (%s, %s, %s, %s, %s)"
    cursor.execute(sql_bottom, (emp_id_bottom, page_num + 1, 'bottom', bottom_half, bottom_pdf_path))
    conn.commit()

# ปิดการเชื่อมต่อและปิดไฟล์ PDF
cursor.close()
conn.close()
pdf_document.close()

print("PDF data processed successfully.")

File: view/admin/function/process_pdf.js
const express = require('express');
const mysql = require('mysql2');
const fs = require('fs');
const pdfLib = require('pdf-lib');  // ใช้ pdf-lib หรือ pdf-parse ตามต้องการ
const pdfParse = require('pdf-parse');
const path = require('path');
const app = express();
const port = 3000;

// เชื่อมต่อกับฐานข้อมูล MySQL
const conn = mysql.createConnection({
    host: 'localhost',
    user: 'root',
    password: '',
    database: 'salary_slip'
});

conn.connect((err) => {
    if (err) throw err;
    console.log('Connected to MySQL database');
});

// ใช้ express.json() เพื่อจัดการ request body
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// API สำหรับประมวลผล PDF
app.post('/process-pdf', async (req, res) => {
    const { filePath, pdfId } = req.body;
    
    if (!filePath || !pdfId) {
        return res.status(400).json({ message: 'Invalid parameters' });
    }

    try {
        const dataBuffer = fs.readFileSync(filePath);  // อ่านไฟล์ PDF
        const pdfData = await pdfParse(dataBuffer);
        const pageCount = pdfData.numpages;

        const permanentFolder = 'permanent_pdfs';
        if (!fs.existsSync(permanentFolder)) {
            fs.mkdirSync(permanentFolder);  // สร้างโฟลเดอร์ถาวรสำหรับจัดเก็บ PDF
        }

        // วนลูปเพื่อประมวลผลแต่ละหน้าใน PDF
        for (let pageNum = 0; pageNum < pageCount; pageNum++) {
            const page = pdfData.pages[pageNum];
            const pageText = page.text;
            const pageHeight = page.height;

            // แบ่งครึ่งบนและครึ่งล่างของหน้า
            const halfIndex = Math.floor(pageText.length / 2);
            const topHalf = pageText.substring(0, halfIndex);
            const bottomHalf = pageText.substring(halfIndex);

            // สร้างไฟล์ PDF สำหรับครึ่งบนและครึ่งล่าง
            const topPdfPath = path.join(permanentFolder, `top_${pageNum + 1}.pdf`);
            const bottomPdfPath = path.join(permanentFolder, `bottom_${pageNum + 1}.pdf`);

            // บันทึกครึ่งบนและครึ่งล่างลงในไฟล์ PDF
            fs.writeFileSync(topPdfPath, topHalf);
            fs.writeFileSync(bottomPdfPath, bottomHalf);

            // ฟังก์ชันสำหรับดึงข้อมูลพนักงาน
            const extractEmpInfo = (text) => {
                const empIdMatch = text.match(/\bP\d{4,}\b/);
                const empId = empIdMatch ? empIdMatch[0] : 'unknown';

                const empNameMatch = text.match(/P\d{4,}\s*(?:ชื่อสกุล|Name|Emp)?[:\-]?\s*([^\n]*)/i);
                const empName = empNameMatch ? empNameMatch[1].trim() : 'unknown';

                const empDepMatch = text.match(/Dep\.\s*(\S.*)/);
                const empDep = empDepMatch ? empDepMatch[1].trim() : 'unknown';

                return { empId, empName, empDep };
            };

            // ฟังก์ชันสำหรับดึงข้อมูลเงินเดือน
            const extractSalaryDetails = (text) => {
                const salaryIncomeMatch = text.match(/รวมเงินได้\s*[:\-]?\s*([\d,]+\.\d{2})/);
                const salaryIncome = salaryIncomeMatch ? parseFloat(salaryIncomeMatch[1].replace(',', '')) : 0.0;

                const salaryDeductionMatch = text.match(/รวมเงินหัก\s*[:\-]?\s*([\d,]+\.\d{2})/);
                const salaryDeduction = salaryDeductionMatch ? parseFloat(salaryDeductionMatch[1].replace(',', '')) : 0.0;

                const salaryNetMatch = text.match(/เงินได้สุทธิ\s*[:\-]?\s*([\d,]+\.\d{2})/);
                const salaryNet = salaryNetMatch ? parseFloat(salaryNetMatch[1].replace(',', '')) : 0.0;

                return { salaryIncome, salaryDeduction, salaryNet };
            };

            // ดึงข้อมูลจากครึ่งบน
            const { empId: empIdTop, empName: empNameTop, empDep: empDepTop } = extractEmpInfo(topHalf);
            const { salaryIncome: salaryIncomeTop, salaryDeduction: salaryDeductionTop, salaryNet: salaryNetTop } = extractSalaryDetails(topHalf);

            // ดึงข้อมูลจากครึ่งล่าง
            const { empId: empIdBottom, empName: empNameBottom, empDep: empDepBottom } = extractEmpInfo(bottomHalf);
            const { salaryIncome: salaryIncomeBottom, salaryDeduction: salaryDeductionBottom, salaryNet: salaryNetBottom } = extractSalaryDetails(bottomHalf);

            // บันทึกข้อมูลครึ่งบนในฐานข้อมูล
            const sqlTop = `
                INSERT INTO tb_payslips (emp_id, emp_name, emp_dep, salary_income, salary_deduction, salary_net, page_number, part, content, slip_divide)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            `;
            conn.query(sqlTop, [empIdTop, empNameTop, empDepTop, salaryIncomeTop, salaryDeductionTop, salaryNetTop, pageNum + 1, 'top', topHalf, topPdfPath]);

            // บันทึกข้อมูลครึ่งล่างในฐานข้อมูล
            const sqlBottom = `
                INSERT INTO tb_payslips (emp_id, emp_name, emp_dep, salary_income, salary_deduction, salary_net, page_number, part, content, slip_divide)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            `;
            conn.query(sqlBottom, [empIdBottom, empNameBottom, empDepBottom, salaryIncomeBottom, salaryDeductionBottom, salaryNetBottom, pageNum + 1, 'bottom', bottomHalf, bottomPdfPath]);
        }

        // ส่งผลลัพธ์กลับไปยังผู้เรียก
        res.json({ message: 'PDF processed and data inserted successfully', status: 'success' });
    } catch (error) {
        console.error('Error processing PDF:', error);
        res.status(500).json({ message: 'PDF processing failed', status: 'error' });
    }
});

// เริ่มต้นเซิร์ฟเวอร์
app.listen(port, () => {
    console.log(`Server is running on http://localhost:${port}`);
});
